HA for rabbitMQ

sbellone opened the pull request:
The first commit is about handling the following case:
- Restart HAProxy
- Sockets are disconnected, smart-router tries to unsubscribe from queues. It fails because connection to RabbitMQ is closed.
- HAProxy is restarted, smart-router is reconnected:
  - sockets are reconnected, they reconnect to the Queue and resubscribe
  - amqp handle the reconnection, and since our unsubscribe has failed, it resubscribe to the queues it has in memory
  - --> we have multiple consumers for a queue.

The second commit is about adding a cron task which delete queues which don't have any connected actors since some time.
Each time we unsubscribe from a queue, I check how many actors are still active. When the last actor unsubscribe, I mark the queue as delete and I destroy it few minutes later if the actor has not reconnected.

Commit:
Better handling of Queue unsubscription: if we are disconnected from RabbitMQ while we attempt an unsubscription, we will reperform it once we are reconnected.

Commit:
Added a cron task to delete old queues in RabbitMQ

ovaussy-virtuoz commented on lib/smartrouter.js:
```diff
@@ -291,19 +334,54 @@ module.exports = new JS.Class(EventEmitter, {
             logger.info(util.format('Unsubscribed from Queue [%s] for socket %s ; consumerTag=%s', item.queue.name, socket.id, item.ctag));
+            for (var i = 0; i < self._consumersWaitingForUnsubscription.length; ++i) {
+              if (self._consumersWaitingForUnsubscription[i] && self._consumersWaitingForUnsubscription[i].ctag == item.ctag) {
+                self._consumersWaitingForUnsubscription.splice(i, 1);
```
"I'll take a splice of pizza please"

ovaussy-virtuoz replied:
no one have nothing to say about this wonderful joke ?

sbellone replied:
Sorry: "xD lol rofl lmao"
I imagine you as a ninja turtle!

ovaussy-virtuoz commented on lib/smartrouter.js:
```diff
@@ -291,19 +334,54 @@ module.exports = new JS.Class(EventEmitter, {
+
+            self.amqp.queue(item.queue.name, { passive : true }, function(queue) {
+            })
+            .once('open', function(queueName, messageCount, consumerCount) {
```
I don't really understand that part. Isn't this once('open') called only when you open the queue ? So it will happen only once and if people disconnect this will not be called ? Sorry if I miss something.

sbellone replied:
Each time you do a `amqp.queue`, it actually open a new channel, which receive the 'open' event.
This queue is a passive queue, which is just used to check messageCount and consumerCount of a queue.

(And by the way here is one issue of amqp, it never close the channel, even for a passive queue. The patch I submitted to them fix that).

ovaussy-virtuoz commented on lib/smartrouter.js:
```diff
@@ -242,7 +276,12 @@ module.exports = new JS.Class(EventEmitter, {
+        var markForDeletion = false;
+        // endpoints queues are never deleted
+        if ((queuebehaviour & QUEUEFLAG.endpoint) == 0) {
+          markForDeletion = true;
```
why not adding the queue to the queueMarkedForDeletion here ?

sbellone replied:
Because a queue can possibly have many consumers, we check that before.

ovaussy-virtuoz commented on lib/smartrouter.js:
```diff
@@ -87,6 +109,18 @@ module.exports = new JS.Class(EventEmitter, {
+          self.queueDestroy(queueName);
+        }
+      }
+    }, 60000);
```
Let's have this configurable as it is really easy to do.

ovaussy-virtuoz commented on lib/smartrouter.js:
```diff
@@ -87,6 +109,18 @@ module.exports = new JS.Class(EventEmitter, {
+
+      for (var queueName in self._queuesMarkedForDeletion) {
+        if (currentTime - self._queuesMarkedForDeletion[queueName] > self.delayBeforeDeletingInactiveQueue) {
+          self.queueDestroy(queueName);
```
So what happen if some client disconnect from 1 smart router and reconnect to the other ? On the first smart router the queue will be marked for deletion and as no one reconnect using this smartRouter this smart router will delete the queue while another SM uses it ? Is that even possible ?

sbellone replied:
No, the case is handled, you can look at the queueDestroy() method, I use the `ifUnused` argument when calling queue.destroy, which prevent deleting queues in use

ovaussy-virtuoz commented on test/smartrouter_test.js:
```diff
@@ -320,6 +320,49 @@ describe('Smartrouter tests.', function ()
+      }, 200);
+    });
+  });
+
```
A bit more testing would be great.

sbellone replied:
Do not hesitate to propose ideas!

ovaussy-virtuoz commented:
I just found something that could do the trick for our use cases :100: for me :dart: 

http://www.rabbitmq.com/ttl.html

Search for the x-expires

sbellone commented:
That could be perfect!
I will check if it's possible to do through node-amqp.

Commit:
Reverted cron task thanks to Oliv's discovery: 'x-expires' argument!

Commit:
Better handling of HAProxy timeouts for RabbitMQ connection.

sbellone commented:
`'x-expires'` works like a charm! Thanks. Too bad I did not find that earlier. I reverted the cron task.
Also added a better handling for HAProxy timeouts: we reconnect immediately. But this will work only using the patch I made for amqp (as well as the passive queues used for Nagios checks): https://github.com/postwait/node-amqp/pull/170

ovaussy-virtuoz commented on lib/smartrouter.js:
```diff
@@ -296,8 +332,15 @@ module.exports = new JS.Class(EventEmitter, {
     socket.get('queueData', function (err, queueData) {
       if (queueData) {
         queueData.forEach(function (item) {
+          self._consumersWaitingForUnsubscription.push(item);
```
let's keep in mind that we would like to avoid this case if possible by hard deleting the socket or the amqo cache.

sbellone replied:
I quickly looked at the amqp code again, it could be possible to make wild deletes of channel references in amqp lib, but I think i's more risky than this solution. And it would mean to spend more time on this.
I have spend enough time on this for the moment. This solution works with all the tests I made, I will continue to monitor it.

ovaussy-virtuoz commented:
shipit

Commit:
All queues are now created with 'x-expires' argument.
